{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Standard Numbering\n",
    "\n",
    "The standard numbering tool is used to number the residues of a protein sequence. It allows for comparison of different protein sequences by aligning them and numbering the residues in a common reference frame.\n",
    "\n",
    "It can be run in two different modes:\n",
    "\n",
    "1. **Pairwise alignment**: This mode aligns two sequences and numbers the residues in a common reference frame. Here a base sequence is provided and the other sequences are aligned to it.\n",
    "2. **Clustal alignment**: This mode aligns a sequence against a multiple sequence alignment and numbers the residues in a common reference frame. Here a base sequence is provided and the other sequences are aligned to it.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "import sys\n",
    "from loguru import logger\n",
    "\n",
    "from pyeed import Pyeed\n",
    "from pyeed.analysis.mutation_detection import MutationDetection\n",
    "from pyeed.analysis.standard_numbering import StandardNumberingTool\n",
    "\n",
    "logger.remove()\n",
    "level = logger.add(sys.stderr, level=\"WARNING\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ðŸ“¡ Connected to database.\n",
      "All data has been wiped from the database.\n",
      "the connection url is bolt://neo4j:12345678@129.69.129.130:7687\n",
      "Loaded /home/nab/Niklas/pyeed/src/pyeed/model.py\n",
      "Connecting to bolt://neo4j:12345678@129.69.129.130:7687\n",
      "Setting up indexes and constraints...\n",
      "\n",
      "Found model.StrictStructuredNode\n",
      " ! Skipping class model.StrictStructuredNode is abstract\n",
      "Found model.Organism\n",
      " + Creating node unique constraint for taxonomy_id on label Organism for class model.Organism\n",
      "{code: Neo.ClientError.Schema.EquivalentSchemaRuleAlreadyExists} {message: An equivalent constraint already exists, 'Constraint( id=12, name='constraint_unique_Organism_taxonomy_id', type='UNIQUENESS', schema=(:Organism {taxonomy_id}), ownedIndex=5 )'.}\n",
      "Found model.Site\n",
      " + Creating node unique constraint for site_id on label Site for class model.Site\n",
      "{code: Neo.ClientError.Schema.EquivalentSchemaRuleAlreadyExists} {message: An equivalent constraint already exists, 'Constraint( id=14, name='constraint_unique_Site_site_id', type='UNIQUENESS', schema=(:Site {site_id}), ownedIndex=7 )'.}\n",
      "Found model.Region\n",
      " + Creating node unique constraint for region_id on label Region for class model.Region\n",
      "{code: Neo.ClientError.Schema.EquivalentSchemaRuleAlreadyExists} {message: An equivalent constraint already exists, 'Constraint( id=19, name='constraint_unique_Region_region_id', type='UNIQUENESS', schema=(:Region {region_id}), ownedIndex=11 )'.}\n",
      "Found model.CatalyticActivity\n",
      " + Creating node unique constraint for catalytic_id on label CatalyticActivity for class model.CatalyticActivity\n",
      "{code: Neo.ClientError.Schema.EquivalentSchemaRuleAlreadyExists} {message: An equivalent constraint already exists, 'Constraint( id=3, name='constraint_unique_CatalyticActivity_catalytic_id', type='UNIQUENESS', schema=(:CatalyticActivity {catalytic_id}), ownedIndex=15 )'.}\n",
      "Found model.StandardNumbering\n",
      " + Creating node unique constraint for name on label StandardNumbering for class model.StandardNumbering\n",
      "{code: Neo.ClientError.Schema.EquivalentSchemaRuleAlreadyExists} {message: An equivalent constraint already exists, 'Constraint( id=16, name='constraint_unique_StandardNumbering_name', type='UNIQUENESS', schema=(:StandardNumbering {name}), ownedIndex=20 )'.}\n",
      "Found model.GOAnnotation\n",
      " + Creating node unique constraint for go_id on label GOAnnotation for class model.GOAnnotation\n",
      "{code: Neo.ClientError.Schema.EquivalentSchemaRuleAlreadyExists} {message: An equivalent constraint already exists, 'Constraint( id=10, name='constraint_unique_GOAnnotation_go_id', type='UNIQUENESS', schema=(:GOAnnotation {go_id}), ownedIndex=4 )'.}\n",
      "Found model.Protein\n",
      " + Creating node unique constraint for accession_id on label Protein for class model.Protein\n",
      "{code: Neo.ClientError.Schema.EquivalentSchemaRuleAlreadyExists} {message: An equivalent constraint already exists, 'Constraint( id=24, name='constraint_unique_Protein_accession_id', type='UNIQUENESS', schema=(:Protein {accession_id}), ownedIndex=13 )'.}\n",
      " + Creating vector index for embedding on label Protein for class model.Protein\n",
      "{code: Neo.ClientError.Schema.EquivalentSchemaRuleAlreadyExists} {message: An equivalent index already exists, 'Index( id=9, name='vector_index_Protein_embedding', type='VECTOR', schema=(:Protein {embedding}), indexProvider='vector-2.0' )'.}\n",
      "Found model.DNA\n",
      " + Creating node unique constraint for accession_id on label DNA for class model.DNA\n",
      "{code: Neo.ClientError.Schema.EquivalentSchemaRuleAlreadyExists} {message: An equivalent constraint already exists, 'Constraint( id=8, name='constraint_unique_DNA_accession_id', type='UNIQUENESS', schema=(:DNA {accession_id}), ownedIndex=21 )'.}\n",
      " + Creating vector index for embedding on label DNA for class model.DNA\n",
      "{code: Neo.ClientError.Schema.EquivalentSchemaRuleAlreadyExists} {message: An equivalent index already exists, 'Index( id=6, name='vector_index_DNA_embedding', type='VECTOR', schema=(:DNA {embedding}), indexProvider='vector-2.0' )'.}\n",
      "Found model.OntologyObject\n",
      " + Creating node unique constraint for name on label OntologyObject for class model.OntologyObject\n",
      "{code: Neo.ClientError.Schema.EquivalentSchemaRuleAlreadyExists} {message: An equivalent constraint already exists, 'Constraint( id=17, name='constraint_unique_OntologyObject_name', type='UNIQUENESS', schema=(:OntologyObject {name}), ownedIndex=23 )'.}\n",
      "\n",
      "Finished 10 classes.\n",
      "âœ… Databse constraints and indexes set up according to Pyeed Graph Object Model.\n"
     ]
    }
   ],
   "source": [
    "uri = \"bolt://129.69.129.130:7687\"\n",
    "user = \"neo4j\"\n",
    "password = \"12345678\"\n",
    "\n",
    "eedb = Pyeed(uri, user=user, password=password)\n",
    "eedb.db.wipe_database(date=\"2025-03-19\")\n",
    "\n",
    "eedb.db.initialize_db_constraints(user=user, password=password)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "ids = [\"AAM15527.1\", \"AAF05614.1\", \"AFN21551.1\", \"CAA76794.1\", \"AGQ50511.1\"]\n",
    "\n",
    "eedb.fetch_from_primary_db(ids, db=\"ncbi_protein\")\n",
    "eedb.fetch_dna_entries_for_proteins()\n",
    "eedb.create_coding_sequences_regions()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "9e75d0d45b214ded9bae2253574798c2",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Output()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"></pre>\n"
      ],
      "text/plain": []
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sn = StandardNumberingTool(name=\"test_standard_numbering_pairwise\")\n",
    "\n",
    "\n",
    "sn.apply_standard_numbering_pairwise(\n",
    "    base_sequence_id=\"AAM15527.1\", db=eedb.db, list_of_seq_ids=ids[0:5]\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "0bbf1bf2c7454c6e9b224de596b51835",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Output()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"></pre>\n"
      ],
      "text/plain": []
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sn.apply_standard_numbering_pairwise(\n",
    "    base_sequence_id=\"AAM15527.1\", db=eedb.db, list_of_seq_ids=ids\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "sn_clustal = StandardNumberingTool(name=\"test_standard_numbering_clustal\")\n",
    "\n",
    "sn_clustal.apply_standard_numbering(\n",
    "    base_sequence_id=\"AAM15527.1\", db=eedb.db, list_of_seq_ids=ids\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "sn_dna = StandardNumberingTool(name=\"test_standard_numbering_dna\")\n",
    "\n",
    "sn_dna.apply_standard_numbering(\n",
    "    base_sequence_id=\"AF190695.1\", db=eedb.db, node_type=\"DNA\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "1269c760e0f743b2937ce95400d3240b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Output()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"></pre>\n"
      ],
      "text/plain": []
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sn_dna_pairwise = StandardNumberingTool(name=\"test_standard_numbering_dna_pairwise\")\n",
    "\n",
    "sn_dna_pairwise.apply_standard_numbering_pairwise(\n",
    "    base_sequence_id=\"AF190695.1\", db=eedb.db, node_type=\"DNA\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "9484ba5fa2cb4d89b8371b0554d54510",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Output()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"></pre>\n"
      ],
      "text/plain": []
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sn_dna_region = StandardNumberingTool(name=\"test_standard_numbering_dna_pairwise_region\")\n",
    "\n",
    "sn_dna_region.apply_standard_numbering_pairwise(\n",
    "    base_sequence_id=\"AF190695.1\", db=eedb.db, node_type=\"DNA\", region_based_sequence='coding sequence'\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In both cases, there are now standard numbering nodes to all the proteins and they have on their edge the standradnumbering data."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pyeed_niklas",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
